src/agents/supervisor/agent.py

```diff
@@ -1,128 +1,115 @@
+from langchain_core.output_parsers.openai_functions import JsonOutputFunctionsParser
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
 
-from utils.agents import create_agent
+from agents.common.state import AgentState, Plan
+from agents.common.utils import filter_messages
 from utils.logging import get_logger
+from utils.models import Model
 
-logger = get_logger(__name__)
-
-langfuse_handler = CallbackHandler(
-    secret_key=os.getenv("LANGFUSE_SECRET_KEY"),
-    public_key=os.getenv("LANGFUSE_PUBLIC_KEY"),
-    host=os.getenv("LANGFUSE_HOST"),
-)
-
-
-class Message(BaseModel):
-    """Message data model."""
-
-    input: str
-    session_id: str
-
-
-class Agent(Protocol):
-    """Agent interface."""
-
-    async def astream(self, message: Message) -> dict[str, Any]:
-        """Stream the input to the supervisor asynchronously."""
-        ...
-
-
-@tool
-def search(query: str) -> list[str]:
-    """Call to get information about Kyma."""
-    # This is a placeholder for the actual implementation
-    return [
-        "Kyma is an opinionated set of Kubernetes-based modular building blocks, "
-        "including all necessary capabilities to develop and run "
-        "enterprise-grade cloud-native applications."
-    ]
-
-
-def should_continue(state: MessagesState) -> Literal["action", "__end__"]:
-    """Return the next node to execute."""
-    last_message = state["messages"][-1]
-    return "action" if last_message.tool_calls else "__end__"
+SUPERVISOR = "Supervisor"
+COMMON = "Common"
+FINALIZER = "Finalize"
 
-
-def filter_messages(messages: list) -> list:
-    """This is very simple helper function which only ever uses the last four messages"""
-    return messages[-10:]
-
-
-def agent_node(
-    state: dict[str, Any], agent: AgentExecutor, name: str
-) -> dict[str, Any]:
-    """It filters the messages and invokes the agent."""
-    state["messages"] = filter_messages(state["messages"])
-
-    result = agent.invoke(state)
-    return {"messages": [HumanMessage(content=result["output"], name=name)]}
+logger = get_logger(__name__)
 
 
 class SupervisorAgent:
-    """Supervisor agent.
-    Currently, it has just prelimiary implementation.
-    I'll be adding more features to it in the future.
-    """
-
-    llm = None
-    memory: BaseCheckpointSaver
-    kyma_agent = None
-    tools = None
-
-    def __init__(self, llm: ChatOpenAI, memory: BaseCheckpointSaver):
-        self.llm = llm
-        self.memory = memory
-        self.tools = [search]
-        self.kyma_agent = functools.partial(
-            agent_node,
-            agent=create_agent(
-                llm,
-                self.tools,
-                "You are Kyma expert. You assist users with Kyma related questions.",
-            ),
-            name="KymaAgent",
+    """Supervisor agent class."""
+
+    model: Model
+    _name: str = SUPERVISOR
+    parser = PydanticOutputParser(pydantic_object=Plan)  # type: ignore
```

Comment:
Is this parser used anywhere?

Reply:
yes, below, you can search for it with `state.subtasks = self.parser.parse(plan.content).subtasks`